# Results files

The pipeline line generates lots of files of which only some are kept. 
These final output files are listed and shortly described in the table below. 
`{sample}` is replaced with the sample name and `{type}` is replaced with the sample type (T or N, where T=Tumor, N=Normal).

## Main output files

| **File**                                                                  | **File type**  | **Description**                                                                   |
|---------------------------------------------------------------------------|----------------|-----------------------------------------------------------------------------------|
| **Alignment**                                                             | :~~:           | ~~                                                                                |
| `results/whatshap_haplotag/{sample}_T.haplotagged.bam`                    | BAM            | mapped haplotagged reads                                                          |
| `results/whatshap_haplotag/{sample}_T.haplotagged.bam.bai`                | BAM index      | index for haplotagged reads                                                       |
| **SNV and INDELs**                                                        | :~~:           | ~~                                                                                |
| `results/whatshap_phase/{sample}_T.phased.include.panel.vcf.gz`           | compressed VCF | phased SNV and InDels called by DeepSomatic                                       |
| **CNVs**                                                                  | :~~:           | ~~                                                                                |
| `results/pbsv_call/{sample}_T.vcf.gz`                                     | compressed VCF | CNVs called by pbsv                                                               |
| `results/hificnv/{sample}_T.vcf.gz`                                       | compressed VCF | CNVs called by HiFiCNV                                                            |
| `results/sniffles2_call/{sample}_T.vcf`                                   | VCF            | post-processed SV called by Sniffles2 that passed quality checks                  |
| `results/sniffles2_call/{sample}_T.snf`                                   | SNF            | unfiltered SV candidates and genome-wide coverage information called by Sniffles2 |
| `results/cnvkit_batch/{sample}/{sample}_T.haplotagged.call.cns`           | CNS            | bin-level log2 ratios, generated by CNVkit                                        |
| `results/cnvkit_batch/{sample}/{sample}_T.haplotagged.targetcoverage.cnn` | CNN            | target and antitarget bin-level coverage, generated by CNVkit                     |
| **Depth of coverage**                                                     | :~~:           | ~~                                                                                |
| `results/mosdepth_bed/{sample}_T.regions.bed.gz`                          | compressed BED | median depth of coverage for each target, generated by mosdepth                   |
| `results/mosdepth_bed/{sample}_T.per-base.bed.gz`                         | compressed BED | per-base depth of coverage for each target, generated by mosdepth                 |
| **Reports**                                                               | :~~:           | ~~                                                                                |
| `results/cnv_html_report/{sample}_T.pathology.cnv_report.html`            | HTML           | CNV report generated by cnvkit                                                    |